app/recurring_job.go

```diff
@@ -466,6 +488,10 @@ func (job *Job) listSnapshotNamesForCleanup(snapshots []longhornclient.Snapshot,
 	return snapshotsToNames(filterSnapshotsNotInTargets(snapshots, retainingSnapshots))
 }
 
+func (job *Job) listSnapshotNamesToCleanupForTaskSnapshotDelete(snapshots []longhornclient.Snapshot) []string {
+	return filterExpiredItems(snapshotsToNameWithTimestamps(snapshots), job.retain+1)
```

Comment:
Why `job.retain+1`?

Reply:
In the current implementation recurring job creates a snapshot and then does the cleanup. For the cleanup task and delete task there will be no snapshot creation. So I assume `+1` should be added for the volume-head.

Reply:
Will Longhorn count the volume head as a snapshot here...?

Reply:
My misunderstanding, you are right Longhorn doesn't count the volume head here. However, `filterExpiredItems` is expecting a new snapshot created before the the clean up. Therefore the `+1` is required to have the correct number of snapshots in the returning list.

Reply:
Sorry...? I do not get the idea...
If there is 1 existing snapshot only and retain is 0, this would be `filterExpiredItems(snapshotsToNameWithTimestamps([snap0]), 1)`. And the returned list is an empty one, which is unexpected...

Reply:
IIUC, if there is 1 existing snapshot only and retain is 0, the input snapshots will also include the volume-head. This will be
`filterExpiredItems(snapshotsToNameWithTimestamps([snap0, volume-head]), 1)`

However, I've missed the retain correction [here](https://github.com/longhorn/longhorn-manager/blob/master/app/recurring_job.go#L235-L238) which returned an empty list when retain value is 0.

Thank you for pointing this out, fixed. 

Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2977/

Reply:
I've already removed the retain correction and filtered in `snapshotsToNameWithTimestamps` by adding a new function argument `excludeVolumeHead`. I think this way will have minimal impact.